File: frontend/src/services/vault.ts
import { apiRequest } from "./api";

export type VaultCredential = {
  id: string;
  site_name: string;
  site_url: string;
  masked_username: string;
  created_at: string;
  last_used_at: string | null;
};

export type VaultListResponse = {
  items: VaultCredential[];
};

export type VaultStorePayload = {
  site_name: string;
  site_url: string;
  username: string;
  password: string;
};

export async function listCredentials(): Promise<VaultListResponse> {
  return apiRequest<VaultListResponse>("/api/v1/vault/credentials");
}

export async function storeCredential(payload: VaultStorePayload): Promise<VaultCredential> {
  return apiRequest<VaultCredential>("/api/v1/vault/credentials", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function deleteCredential(credentialId: string): Promise<{ deleted: boolean }> {
  return apiRequest<{ deleted: boolean }>(`/api/v1/vault/credentials/${credentialId}`, {
    method: "DELETE",
  });
}

File: backend/app/schemas/vault.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class CredentialStorePayload(BaseModel):
    site_name: str = Field(min_length=2, max_length=100)
    site_url: str = Field(min_length=8, max_length=255)
    username: str = Field(min_length=3, max_length=255)
    password: str = Field(min_length=8, max_length=255)


class VaultCredentialItem(BaseModel):
    id: str
    site_name: str
    site_url: str
    masked_username: str
    created_at: datetime
    last_used_at: datetime | None


class VaultCredentialListData(BaseModel):
    items: list[VaultCredentialItem]


class DeleteCredentialData(BaseModel):
    deleted: bool
